interop-tests/src/lib.rs

```diff
@@ -0,0 +1,319 @@
+}
+
+// Map each role to its current key.
+type RoleKeys = HashMap<&'static str, Ed25519PrivateKey>;
+
+fn init_role_keys(json_keys: &TestKeys) -> RoleKeys {
+    let mut keys = HashMap::new();
+    keys.insert("root", json_keys.root[0][0].to_private_key());
+    keys.insert("snapshot", json_keys.snapshot[0][0].to_private_key());
+    keys.insert("targets", json_keys.targets[0][0].to_private_key());
+    keys.insert("timestamp", json_keys.timestamp[0][0].to_private_key());
+    keys
+}
+
+fn copy_repo(dir: &Path, step: u8) {
+    let src = Path::new(dir)
+        .join((step - 1).to_string())
+        .join("repository");
+    let dst = Path::new(dir).join(step.to_string()).join("repository");
+
+    for (path, f) in read_dir_files(&src) {
+        let path = dst.join(path);
+        if let Some(parent) = path.parent() {
+            if !parent.exists() {
+                fs::create_dir_all(parent).unwrap();
+            }
+        }
+        fs::write(path, &f).unwrap();
+    }
+}
+
+pub fn read_dir_files(path: &Path) -> BTreeMap<PathBuf, String> {
+    let mut entries = BTreeMap::<PathBuf, _>::new();
+    for entry in WalkDir::new(path) {
+        let entry = entry.unwrap();
+        if entry.metadata().unwrap().is_file() {
+            let f = fs::read_to_string(entry.path()).unwrap();
+
+            // Strip off the path prefix.
+            let path = entry.into_path().strip_prefix(&path).unwrap().to_path_buf();
+
+            entries.insert(path, f);
+        }
+    }
+
+    entries
+}
+
+// updates the root metadata. If root_signer is Some, use that to sign the
+// metadata, otherwise use keys["root"].
+async fn update_root(
+    repo: &FileSystemRepository<JsonPretty>,
+    keys: &RoleKeys,
+    root_signer: Option<&dyn PrivateKey>,
+    version: u32,
+    consistent_snapshot: bool,
+) {
+    let signer = match root_signer {
+        Some(k) => k,
+        None => keys.get("root").unwrap(),
+    };
+
+    // Same expiration as go-tuf metadata generator.
+    let expiration = Utc.ymd(2100, 1, 1).and_hms(0, 0, 0);
+
+    let mut root = RootMetadataBuilder::new()
+        .root_key(keys.get("root").unwrap().public().clone())
+        .expires(expiration)
+        .version(version)
+        .snapshot_key(keys.get("snapshot").unwrap().public().clone())
+        .targets_key(keys.get("targets").unwrap().public().clone())
+        .timestamp_key(keys.get("timestamp").unwrap().public().clone())
+        .consistent_snapshot(consistent_snapshot)
+        .signed::<JsonPretty>(signer)
+        .unwrap();
+
+    // If we rotated the root, sign it again with the new key.
+    if root_signer.is_some() {
+        root.add_signature(keys.get("root").unwrap()).unwrap()
+    };
+
+    let root_path = MetadataPath::from_role(&Role::Root);
+    repo.store_metadata(
+        &root_path,
+        &MetadataVersion::Number(version),
+        &mut root.to_raw().unwrap().as_bytes(),
+    )
+    .await
+    .unwrap();
+    repo.store_metadata(
```

Comment:
I know why you have to do this twice (to get both a versioned and unversioned root metadata), but a comment would be helpful as well, since it looks weird in review.

Reply:
Added a comment.